src/auth.ts

```diff
@@ -1,5 +1,182 @@
-import { AuthOptions, Authentication } from "./types";
+import {
+  oauthAuthorizationUrl,
+  OAuthAppOptions,
+  GitHubAppOptions,
+} from "@octokit/oauth-authorization-url";
 
-export async function auth(options: AuthOptions): Promise<Authentication> {
-  // TODO: add implementation
+import errors from "./errors";
+import {
+  ClientType,
+  ExpirationSetting,
+  State,
+  Authentication,
+  AuthCommands,
+  OAuthApp,
+} from "./types";
+
+type GetWebFlowAuthorizationUrlOptions<Client extends ClientType> =
+  Client extends OAuthApp ? OAuthAppOptions : GitHubAppOptions;
+
+export async function auth<
+  Client extends ClientType,
+  Expiration extends ExpirationSetting
+>(
+  state: State<Client, Expiration>,
+  options: AuthCommands<Expiration> = { type: "getSession" }
+): Promise<Authentication<Client, Expiration> | null> {
+  switch (options.type) {
+    case "getSession": {
+      // Unless both `code` and `state` search parameters are present, get
+      // authentication from internal state, fail over to authentication store.
+      const searchParams = new URLSearchParams(location.search);
+      if (!searchParams.get("code") || !searchParams.get("state")) {
+        if (state.authentication) return state.authentication;
+        if (!state.authStore) return null;
+        return (state.authentication = await state.authStore.get());
+      }
+    }
+
+    // Exchange `code` search parameter for an OAuth authentication.
+    case "createToken": {
+      // Both `code` and `state` should present.
+      const url = new URL(location.href);
+      const code = url.searchParams.get("code");
+      const receivedState = url.searchParams.get("state");
+      if (!code || !receivedState) throw errors.codeOrStateMissing;
+
+      // Received `state` must match provided.
+      if (state.stateStore) {
+        const providedState = await state.stateStore.get();
+        await state.stateStore.del();
+        if (receivedState !== providedState) throw errors.stateMismatch;
+      }
+
+      // Remove `code` and `state` query parameters from url.
+      url.searchParams.delete("code");
+      url.searchParams.delete("state");
+      const redirectUrl = url.href;
+      history.pushState({}, "", redirectUrl);
+
+      // Exchange `code` search parameter for an OAuth authentication.
+      const route = state.serviceHost + state.pathPrefix + "/token";
+      const { data } = await state.request(route, {
+        method: "POST",
+        code,
+        redirectUrl,
+        state: receivedState,
+        // TODO: I’m not sure why `oauth-app.js` asks the `state`. GitHub API
+        // does not accept it as documented: https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#parameters-1.
+        // This should be unnecessary: https://github.com/octokit/oauth-methods.js/blob/main/src/exchange-web-flow-code.ts#L78
```

Comment:
It used to be documented, but I brought it up with support and turns out it was never used in the code, so it got removed. You can remove the `state` parameter here, I'll make a follow up issue to do the same in https://github.com/octokit/oauth-methods.js/#exchangewebflowcode